client/my-sites/checkout/checkout/billing-details-form-fields.jsx

```diff
@@ -0,0 +1,591 @@
+						position={ isMobile() ? 'left' : 'right' }
+					>
+						{ translate(
+							'Protects your identity and prevents spam by keeping your contact information off the Internet.'
+						) }
+					</InfoPopover>
+				</span>
+			</label>
+		);
+	}
+
+	renderDomainDetailsFields() {
+		const { translate } = this.props;
+		const { showDomainContactDetails } = this.state;
+
+		if ( ! this.needsDomainDetails() ) {
+			return false;
+		}
+
+		return (
+			<div className="checkout__domain-registration-information">
+				<div className="checkout__domain-registration-information-title">
+					{ translate( 'Domain Registration Information' ) }
+				</div>
+				{ this.renderDomainPrivacy() }
+				<label className="checkout__different-domain-details-label">
+					<FormCheckbox
+						defaultChecked={ ! showDomainContactDetails }
+						className="checkout__different-domain-details"
+						onChange={ this.handleDomainDetailsChange.bind( this ) }
+					/>
+					<span>
+						{ translate( 'Use this as my domain contact information' ) }
+						<InfoPopover
+							className="checkout__item-tip-info"
+							position={ isMobile() ? 'left' : 'right' }
+						>
+							{ translate(
+								'Domain owners are required to share contact information in a public database of all domains.'
+							) }
+						</InfoPopover>
+					</span>
+				</label>
+			</div>
+		);
+	}
+
+	renderGAppsFieldset() {
+		const countryCode = this.getCountryCode();
+		return (
+			<div className="contact-details-form-fields__g-apps g-apps-fieldset">
+				<CountrySelect
+					label={ this.props.translate( 'Country' ) }
+					countriesList={ this.props.countriesList }
+					{ ...this.getFieldProps( 'country-code', true ) }
+				/>
+
+				<Input
+					label={ getPostCodeLabelText( countryCode ) }
+					{ ...this.getFieldProps( 'postal-code', true ) }
+				/>
+			</div>
+		);
+	}
+
+	render() {
+		const { translate } = this.props;
+		const { showDomainContactDetails } = this.state;
+
+		return (
+			<div>
+				<FormFieldset className="contact-details-form-fields">
+					{ this.createField( 'first-name', Input, {
+						label: translate( 'First Name' ),
+					} ) }
+
+					{ this.createField( 'last-name', Input, {
+						label: translate( 'Last Name' ),
+					} ) }
+
+					{ this.props.needsOnlyGoogleAppsDetails
+						? this.renderGAppsFieldset()
+						: this.renderContactDetailsFields() }
+
+					<div className="contact-details-form-fields__extra-fields">{ this.props.children }</div>
+
+					{ this.renderDomainDetailsFields() }
+
+					<QueryDomainCountries />
+				</FormFieldset>
+				{ showDomainContactDetails ? <DomainDetailsForm /> : false }
+			</div>
+		);
+	}
+}
+
+export default connect( ( state, props ) => {
+	const contactDetails = props.contactDetails;
+	const countryCode = contactDetails.countryCode;
+
+	const hasCountryStates =
+		contactDetails && contactDetails.countryCode
+			? ! isEmpty( getCountryStates( state, contactDetails.countryCode ) )
+			: false;
```

Comment:
It equals to `const hasCountryStates = countryCode && ! isEmpty( getCountryStates( state, countryCode ))`.